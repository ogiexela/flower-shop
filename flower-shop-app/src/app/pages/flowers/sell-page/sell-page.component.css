.new-flower-form {
    align-content: center;
    align-items: center;
    align-self: center;
    display: flex;
    flex-direction: column;
    justify-content: center;
    width: 500px;
}

.form-input {
    display: flex;
    flex-direction: column;
    margin: 10px;
    width: 100%;
}

.ng-valid[required], .ng-valid.required {
    border-left: 5px solid #42A948;
}

.ng-invalid:not(form) {
    border-left: 5px solid #a94442;
}

.form-controls {
    align-items: center;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap-reverse;
    justify-content: center;
    margin: 20px;
}

.form-controls button {
    cursor: pointer;
    margin: 10px;
}

@media only screen and (max-width: 600px) {
    .new-flower-form {
        max-width: 100%;
        width: 100%;
    }

    .form-controls {
        align-items: stretch;
        flex-direction: column;
        flex-grow: 1;
        flex-wrap: wrap-reverse;
        justify-content: stretch;
    }
}