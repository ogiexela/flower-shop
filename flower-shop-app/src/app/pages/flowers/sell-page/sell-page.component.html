<form
    #flowerForm="ngForm"
    (ngSubmit)="onSubmit()"
    class="new-flower-form"
>
    <h1>New Flower</h1>
    <div class="form-input">
        <label>SKU</label>
        <input
            [(ngModel)]="newFlower.sku"
            class="form-control"
            id="flowerSku"
            name="sku"
            required
            type="text"
        >
    </div>
    <div class="form-input">
        <label>Name</label>
        <input
            [(ngModel)]="newFlower.name"
            class="form-control"
            id="flowerName"
            name="name"
            required
            type="text"
        >
    </div>
    <div class="form-input">
        <label>Image</label>
        <input
            [(ngModel)]="newFlower.image"
            class="form-control"
            id="flowerImage"
            name="image"
            required
            type="text"
        >
    </div>
    <div class="form-input">
        <label>Price</label>
        <input
            [(ngModel)]="newFlower.price"
            class="form-control"
            id="flowerPrice"
            name="price"
            required
            type="number"
        >
    </div>
    <div class="form-input">
        <label>Rating</label>
        <input
            [(ngModel)]="newFlower.rating"
            class="form-control"
            id="flowerRating"
            max="5"
            min="0"
            name="rating"
            type="number"
        >
    </div>
    <div class="form-controls">
        <button
            [disabled]="!flowerForm.form.valid"
            class="btn btn-success"
            id="submit_flower"
            type="submit"
        >Submit</button>
        <button
            (click)="clearFlower(); flowerForm.reset()"
            class="btn btn-default"
            id="reset_flower"
            type="button"
        >Reset</button>
    </div>
</form>
