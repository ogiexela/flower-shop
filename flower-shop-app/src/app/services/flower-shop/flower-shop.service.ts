import { HttpClient } from '@angular/common/http';
import {
  Inject,
  Injectable,
} from '@angular/core';
import { Flower } from '../../models/flower.model';
import { Flowers } from '../../models/flowers.model';

@Injectable({
  providedIn: 'root'
})
export class FlowerShopService {

  private readonly API_URL = '/api';

  constructor(@Inject(HttpClient) private http: HttpClient) { }

  public getFlowers(page: number = 1, offset: number = 0, limit: number = 25) {
    return this.http.get<Flowers>(`${this.API_URL}/flower?limit=${limit}&offset=${offset}&page=${page}`);
  }

  public createFlower(flower: Flower) {
    return this.http.post<Flower>(`${this.API_URL}/flower`, flower);
  }
}
