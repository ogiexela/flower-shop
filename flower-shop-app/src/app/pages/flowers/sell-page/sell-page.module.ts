import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { FlowerModule } from '../../../components/flower/flower.module';
import { SellPageComponent } from './sell-page.component';

@NgModule({
  declarations: [
      SellPageComponent,
  ],
  imports: [
    CommonModule,
    FlowerModule,
    FormsModule
  ]
})
export class SellPageModule { }
