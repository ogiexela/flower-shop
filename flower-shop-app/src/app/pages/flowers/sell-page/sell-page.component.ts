import { take } from 'rxjs/operators';
import {
  Component,
  Inject,
  ViewChild,
} from '@angular/core';
import { Flower } from '../../../models/flower.model';
import { FlowerShopService } from '../../../services/flower-shop/flower-shop.service';

@Component({
  selector: 'app-sell-page',
  styleUrls: ['./sell-page.component.css'],
  templateUrl: './sell-page.component.html',
})
export class SellPageComponent {

  @ViewChild('flowerForm', { static: true}) private form: HTMLFormElement;
  public newFlower: Flower = {};

  constructor(
    @Inject(FlowerShopService) private readonly flowerShopServer: FlowerShopService,
  ) { }

  public onSubmit(): void {
    this.flowerShopServer
      .createFlower(this.newFlower).pipe(take(1))
      .subscribe(
        (createdFlower) => {
          alert(`Created new flower '${createdFlower.name}'.`);
          this.newFlower = {};
          this.form.reset();
        },
        () => {
          alert('Failed to create new flower.');
        });
  }

  public clearFlower(): void {
    this.newFlower = {};
  }

}
