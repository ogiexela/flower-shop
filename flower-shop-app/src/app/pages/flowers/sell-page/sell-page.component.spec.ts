import { HttpClientTestingModule } from '@angular/common/http/testing';
import {
  async,
  ComponentFixture,
  TestBed,
} from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { FlowerShopService } from '../../../services/flower-shop/flower-shop.service';
import { SellPageComponent } from './sell-page.component';
import { SellPageModule } from './sell-page.module';

class MockFlowerShopServiceService {
  public createFlower(newFlower) {
    return {
      pipe: () => {
        return {
          subscribe: (next, error) => {
            next(newFlower);
            error();
          }
        };
      }
    };
  }
}

describe('SellPageComponent', () => {
  let component: SellPageComponent;
  let fixture: ComponentFixture<SellPageComponent>;

  beforeEach(async(() => {
    window.alert = jest.fn();

    TestBed.configureTestingModule({
      declarations: [],
      imports: [SellPageModule, HttpClientTestingModule],
      providers: [
        HttpClientTestingModule,
        { provide: FlowerShopService, useClass: MockFlowerShopServiceService }
      ]
    })
      .compileComponents()
      .then(() => {
        fixture = TestBed.createComponent(SellPageComponent);
        fixture.detectChanges();
        component = fixture.componentInstance;
      })
      .catch(error => console.error(error));
  }));

  it('should create', () => {
    fixture.detectChanges();
    expect(fixture.nativeElement).toMatchSnapshot('empty_state');
  });

  it('submit button is disabled when form is invalid', () => {
    fixture.detectChanges();

    component.onSubmit = jest.fn();

    const submitButton = fixture.debugElement.query(By.css('#submit_flower'));
    submitButton.nativeElement.click();

    expect(component.onSubmit).toBeCalledTimes(0);
  });

  it('able to clear form', () => {
    fixture.detectChanges();

    const hostElement = fixture.nativeElement;
    const flowerSkuInput: HTMLInputElement = hostElement.querySelector('#flowerSku');
    const flowerName: HTMLInputElement = hostElement.querySelector('#flowerName');
    const flowerImage: HTMLInputElement = hostElement.querySelector('#flowerImage');
    const flowerPrice: HTMLInputElement = hostElement.querySelector('#flowerPrice');
    const flowerRating: HTMLInputElement = hostElement.querySelector('#flowerRating');

    const testData = {
      sku: 'f1',
      name: 'flower1',
      image: 'flower_f1.jpg',
      price: '4.99',
      rating: '3',
    };

    flowerSkuInput.value = testData.sku;
    flowerSkuInput.dispatchEvent(new Event('input'));

    flowerName.value = testData.name;
    flowerName.dispatchEvent(new Event('input'));

    flowerImage.value = testData.image;
    flowerImage.dispatchEvent(new Event('input'));

    flowerPrice.value = testData.price;
    flowerPrice.dispatchEvent(new Event('input'));

    flowerRating.value = testData.rating;
    flowerRating.dispatchEvent(new Event('input'));

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot('full_form_valid_state');

    expect(flowerSkuInput.value).toBe(testData.sku);
    expect(flowerName.value).toBe(testData.name);
    expect(flowerImage.value).toBe(testData.image);
    expect(flowerPrice.value).toBe(testData.price);
    expect(flowerRating.value).toBe(testData.rating);

    const resetButton = fixture.debugElement.query(By.css('#reset_flower'));
    resetButton.nativeElement.click();

    fixture.detectChanges();

    expect(flowerSkuInput.value).toBe('');
    expect(flowerName.value).toBe('');
    expect(flowerImage.value).toBe('');
    expect(flowerPrice.value).toBe('');
    expect(flowerRating.value).toBe('');

    expect(fixture.nativeElement).toMatchSnapshot('empy_form_invalid_state');
  });

  it('able to submit valid form', () => {
    fixture.detectChanges();

    const hostElement = fixture.nativeElement;
    const flowerSkuInput: HTMLInputElement = hostElement.querySelector('#flowerSku');
    const flowerName: HTMLInputElement = hostElement.querySelector('#flowerName');
    const flowerImage: HTMLInputElement = hostElement.querySelector('#flowerImage');
    const flowerPrice: HTMLInputElement = hostElement.querySelector('#flowerPrice');
    const flowerRating: HTMLInputElement = hostElement.querySelector('#flowerRating');

    const testData = {
      sku: 'f1',
      name: 'flower1',
      image: 'flower_f1.jpg',
      price: '4.99',
      rating: '3',
    };

    flowerSkuInput.value = testData.sku;
    flowerSkuInput.dispatchEvent(new Event('input'));

    flowerName.value = testData.name;
    flowerName.dispatchEvent(new Event('input'));

    flowerImage.value = testData.image;
    flowerImage.dispatchEvent(new Event('input'));

    flowerPrice.value = testData.price;
    flowerPrice.dispatchEvent(new Event('input'));

    flowerRating.value = testData.rating;
    flowerRating.dispatchEvent(new Event('input'));

    fixture.detectChanges();

    const flowerShopService = TestBed.get(FlowerShopService);

    const createFlowerSpy = jest.spyOn(flowerShopService, 'createFlower');

    const submitButton = fixture.debugElement.query(By.css('#submit_flower'));
    submitButton.nativeElement.click();

    expect(createFlowerSpy.mock.calls).toMatchSnapshot('create_flower_call');

  });

  it('form is cleared after flower is created', async () => {
    fixture.detectChanges();

    const hostElement = fixture.nativeElement;
    const flowerSkuInput: HTMLInputElement = hostElement.querySelector('#flowerSku');
    const flowerName: HTMLInputElement = hostElement.querySelector('#flowerName');
    const flowerImage: HTMLInputElement = hostElement.querySelector('#flowerImage');
    const flowerPrice: HTMLInputElement = hostElement.querySelector('#flowerPrice');
    const flowerRating: HTMLInputElement = hostElement.querySelector('#flowerRating');

    const testData = {
      sku: 'f1',
      name: 'flower1',
      image: 'flower_f1.jpg',
      price: '4.99',
      rating: '3',
    };

    flowerSkuInput.value = testData.sku;
    flowerSkuInput.dispatchEvent(new Event('input'));

    flowerName.value = testData.name;
    flowerName.dispatchEvent(new Event('input'));

    flowerImage.value = testData.image;
    flowerImage.dispatchEvent(new Event('input'));

    flowerPrice.value = testData.price;
    flowerPrice.dispatchEvent(new Event('input'));

    flowerRating.value = testData.rating;
    flowerRating.dispatchEvent(new Event('input'));

    fixture.detectChanges();

    const flowerShopService = TestBed.get(FlowerShopService);

    const createFlowerSpy = jest.spyOn(flowerShopService, 'createFlower');

    const alertSpy = jest.spyOn(window, 'alert');

    const submitButton = fixture.debugElement.query(By.css('#submit_flower'));
    submitButton.nativeElement.click();

    fixture.detectChanges();

    await fixture.whenStable();

    expect(flowerSkuInput.value).toBe('');
    expect(flowerName.value).toBe('');
    expect(flowerImage.value).toBe('');
    expect(flowerPrice.value).toBe('');
    expect(flowerRating.value).toBe('');

    expect(createFlowerSpy.mock.calls).toMatchSnapshot('create_flower_call');
    expect(alertSpy.mock.calls).toMatchSnapshot('alert_calls');
    expect(fixture.nativeElement).toMatchSnapshot('empy_form_invalid_state');
  });
});
