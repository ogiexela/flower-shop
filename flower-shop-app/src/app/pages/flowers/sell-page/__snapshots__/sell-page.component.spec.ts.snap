// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SellPageComponent able to clear form: empy_form_invalid_state 1`] = `
<div
  id="root2"
>
  <form
    class="new-flower-form ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <h1>
      New Flower
    </h1>
    <div
      class="form-input"
    >
      <label>
        SKU
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        id="flowerSku"
        name="sku"
        required=""
        type="text"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Name
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        id="flowerName"
        name="name"
        required=""
        type="text"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Image
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        id="flowerImage"
        name="image"
        required=""
        type="text"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Price
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        id="flowerPrice"
        name="price"
        required=""
        type="number"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Rating
      </label>
      <input
        class="form-control ng-untouched ng-valid ng-pristine"
        id="flowerRating"
        max="5"
        min="0"
        name="rating"
        type="number"
      />
    </div>
    <div
      class="form-controls"
    >
      <button
        class="btn btn-success"
        disabled=""
        id="submit_flower"
        type="submit"
      >
        Submit
      </button>
      <button
        class="btn btn-default"
        id="reset_flower"
        type="button"
      >
        Reset
      </button>
    </div>
  </form>
</div>
`;

exports[`SellPageComponent able to clear form: full_form_valid_state 1`] = `
<div
  id="root2"
>
  <form
    class="new-flower-form ng-untouched ng-dirty ng-valid"
    novalidate=""
  >
    <h1>
      New Flower
    </h1>
    <div
      class="form-input"
    >
      <label>
        SKU
      </label>
      <input
        class="form-control ng-untouched ng-dirty ng-valid"
        id="flowerSku"
        name="sku"
        required=""
        type="text"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Name
      </label>
      <input
        class="form-control ng-untouched ng-dirty ng-valid"
        id="flowerName"
        name="name"
        required=""
        type="text"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Image
      </label>
      <input
        class="form-control ng-untouched ng-dirty ng-valid"
        id="flowerImage"
        name="image"
        required=""
        type="text"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Price
      </label>
      <input
        class="form-control ng-untouched ng-dirty ng-valid"
        id="flowerPrice"
        name="price"
        required=""
        type="number"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Rating
      </label>
      <input
        class="form-control ng-untouched ng-valid ng-dirty"
        id="flowerRating"
        max="5"
        min="0"
        name="rating"
        type="number"
      />
    </div>
    <div
      class="form-controls"
    >
      <button
        class="btn btn-success"
        id="submit_flower"
        type="submit"
      >
        Submit
      </button>
      <button
        class="btn btn-default"
        id="reset_flower"
        type="button"
      >
        Reset
      </button>
    </div>
  </form>
</div>
`;

exports[`SellPageComponent able to submit valid form: create_flower_call 1`] = `
Array [
  Array [
    Object {
      "image": "flower_f1.jpg",
      "name": "flower1",
      "price": 4.99,
      "rating": 3,
      "sku": "f1",
    },
  ],
]
`;

exports[`SellPageComponent form is cleared after flower is created: alert_calls 1`] = `
Array [
  Array [
    "Created new flower 'flower1'.",
  ],
  Array [
    "Failed to create new flower.",
  ],
]
`;

exports[`SellPageComponent form is cleared after flower is created: create_flower_call 1`] = `
Array [
  Array [
    Object {
      "image": "flower_f1.jpg",
      "name": "flower1",
      "price": 4.99,
      "rating": 3,
      "sku": "f1",
    },
  ],
]
`;

exports[`SellPageComponent form is cleared after flower is created: empy_form_invalid_state 1`] = `
<div
  id="root4"
>
  <form
    class="new-flower-form ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <h1>
      New Flower
    </h1>
    <div
      class="form-input"
    >
      <label>
        SKU
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        id="flowerSku"
        name="sku"
        required=""
        type="text"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Name
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        id="flowerName"
        name="name"
        required=""
        type="text"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Image
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        id="flowerImage"
        name="image"
        required=""
        type="text"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Price
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        id="flowerPrice"
        name="price"
        required=""
        type="number"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Rating
      </label>
      <input
        class="form-control ng-untouched ng-valid ng-pristine"
        id="flowerRating"
        max="5"
        min="0"
        name="rating"
        type="number"
      />
    </div>
    <div
      class="form-controls"
    >
      <button
        class="btn btn-success"
        disabled=""
        id="submit_flower"
        type="submit"
      >
        Submit
      </button>
      <button
        class="btn btn-default"
        id="reset_flower"
        type="button"
      >
        Reset
      </button>
    </div>
  </form>
</div>
`;

exports[`SellPageComponent should create: empty_state 1`] = `
<div
  id="root0"
>
  <form
    class="new-flower-form ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <h1>
      New Flower
    </h1>
    <div
      class="form-input"
    >
      <label>
        SKU
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        id="flowerSku"
        name="sku"
        required=""
        type="text"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Name
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        id="flowerName"
        name="name"
        required=""
        type="text"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Image
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        id="flowerImage"
        name="image"
        required=""
        type="text"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Price
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        id="flowerPrice"
        name="price"
        required=""
        type="number"
      />
    </div>
    <div
      class="form-input"
    >
      <label>
        Rating
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-valid"
        id="flowerRating"
        max="5"
        min="0"
        name="rating"
        type="number"
      />
    </div>
    <div
      class="form-controls"
    >
      <button
        class="btn btn-success"
        disabled=""
        id="submit_flower"
        type="submit"
      >
        Submit
      </button>
      <button
        class="btn btn-default"
        id="reset_flower"
        type="button"
      >
        Reset
      </button>
    </div>
  </form>
</div>
`;
